File: day13/13-1.py
import re
from functools import lru_cache

def parse_file(filename):
    with open(filename, 'r') as f:
        content = f.read()
  
    return [
        {
            line[7] if len(line.split(': ')[0]) == 8 else 'P': {'X': int(x), 'Y': int(y)}
            for line in entry.splitlines()
            for x, y in [re.findall(r'\d+', line.split(': ')[1])]
        }
        for entry in content.split('\n\n')
    ]
    
entries = parse_file('13-input')

A_COST, B_COST = 3, 1

def calc_cost(px, py, ax, ay, bx, by) -> int:
    @lru_cache(maxsize=None)
    def calc_cost_this_A_this_B(px, py, a_pressed=0, b_pressed=0):
        # Base cases
        # Success
        if px == 0 and py == 0:
            return a_pressed * A_COST + b_pressed * B_COST
        # Failure: return 0 means spend 0 tokens trying to get this prize, because it's impossible
        if px < 0 or py < 0 or a_pressed > 100 or b_pressed > 100:
            return 0
        
        # Get the minimum, but skip zeroes
        cost_of_pressing_A = calc_cost_this_A_this_B(px - ax, py - ay, a_pressed + 1, b_pressed)
        cost_of_pressing_B = calc_cost_this_A_this_B(px - bx, py - by, a_pressed, b_pressed + 1)
        
        return min(cost_of_pressing_A, cost_of_pressing_B) if cost_of_pressing_A and cost_of_pressing_B else cost_of_pressing_A if cost_of_pressing_A else cost_of_pressing_B
    
    return calc_cost_this_A_this_B(px, py)

print(
    sum(
        (
            calc_cost(
                entry['P']['X'], entry['P']['Y'],
                entry['A']['X'], entry['A']['Y'],
                entry['B']['X'], entry['B']['Y']
            ) for entry in entries
        )
    )
)

File: day13/13-1.c
#include <math.h>
#include <stdio.h>

#define ENTRIES 320
#define A_COST 3
#define B_COST 1
#define EPSILON 0.001

typedef __uint8_t u8;
typedef __uint64_t u64;

u64 calc_cost(u8 ax, u8 ay, u8 bx, u8 by, u64 px, u64 py);
void read_input(u8 *ax, u8 *ay, u8 *bx, u8 *by, u64 *px, u64 *py);

void main(int argc, char const *argv[])
{
    u8 ax[ENTRIES], ay[ENTRIES], bx[ENTRIES], by[ENTRIES];
    u64 px[ENTRIES], py[ENTRIES];
    read_input(ax, ay, bx, by, px, py);

    u64 tokens = 0;
    for (short i = 0; i < ENTRIES; i++) {
        tokens += calc_cost(ax[i], ay[i], bx[i], by[i], px[i], py[i]);
    }
    printf("Tokens necessary: %ld\n", tokens);
}

u64 calc_cost(u8 ax, u8 ay, u8 bx, u8 by, u64 px, u64 py) {
    double denominator = ax*by - bx*ay;
    double a_pressed = (px*by - bx*py) / denominator;
    double b_pressed = (ax*py - ay*px) / denominator;
    u8 impossible = (fabs(a_pressed - round(a_pressed)) > EPSILON)
                 || (fabs(b_pressed - round(b_pressed)) > EPSILON);
    return impossible ? 0 : (round(a_pressed) * A_COST) + (round(b_pressed) * B_COST);
}

void read_input(u8 *ax, u8 *ay, u8 *bx, u8 *by, u64 *px, u64 *py) {
    FILE *fptr = fopen("13-input", "r");
    char buffer[32];
    for(short i = 0; i < ENTRIES; i++) {
        fgets(buffer, 32, fptr);
        sscanf(buffer, "Button A: X+%d, Y+%d", &ax[i], &ay[i]);
        fgets(buffer, 32, fptr);
        sscanf(buffer, "Button B: X+%d, Y+%d", &bx[i], &by[i]);
        fgets(buffer, 32, fptr);
        sscanf(buffer, "Prize: X+%d, Y+%d", &px[i], &py[i]);
        if (i != ENTRIES -1) {
            fgets(buffer, 32, fptr);
        }
    }
}
